pageserver/src/pgdatadir_mapping.rs

```diff
@@ -327,45 +332,49 @@ impl Timeline {
                 )
                 .await?;
 
-            if cmp {
-                high = mid;
-            } else {
+            if !cmp {
+                // We either 1) found commit with timestamp **before** `search_timestamp`;
+                // or 2) we haven't found any commit records at all.
+                // Search with a larger LSN, in case 1) to try to find a more recent commit
+                // (but still **before** target timestamp); and in case 2) to fetch more
+                // SLRU segments for `clog`.
                 low = mid + 1;
+            } else {
+                // We found only more recent commits, search in the older range.
+                high = mid;
             }
         }
+        // If `found_smaller == true`, `low` is the LSN of the first commit record
+        // **before** the `search_timestamp` + 1 (to hit the while loop exit condition).
+        // Substitute 1 to get exactly the commit LSN.
```

Comment:
```suggestion
        // Subtract 1 to get exactly the commit LSN.
```

Reply:
The comment previously said "low is the LSN of the first commit record *after* the search_timestamp". Was that incorrect?

Reply:
First we observe:

1. we exit the while loop only if `low >= high` (there is no break and it's the inversion of the continuation conditional)
2. `high` is only ever reduced to lsns for which `is_latest_commit_timestamp_ge_than` returns true, so for which there are commits that have timestamps that are larger than or equal to the timestamp we are looking for
3. `low > high` never occurs (unless we started out that way), as `mid` is computed in a way that it is always strictly smaller than `low` (as `low < high` must hold for `mid` to be computed, and even if the two are closest, so `low +1 == high`,  `mid` is then `(low + low + 1) / 2`, where `low + low` is an even number so the halving rounds down to `mid = low`), so `mid + 1 <= low`.

We can combine 1 and 3 into:

* when we exit the while loop, `low == high` holds (let's discard us starting out with `low > high` being initialized that way)

Combine this with the invariants established in 2, which now apply for low after the while has terminated. In other words, `low` is an lsn for which `is_latest_commit_timestamp_ge_than` returns true.

And it is the lowest such lsn (that is >= than the start value), as we only adjust `low` upwards if we are sure that all values `< low` have commits with timestamps strictly smaller than the search timestamp.

So yes, what the original comment said is right, and it should read "low is the LSN of the first commit record after (or at) the search_timestamp".

Reply:
So basically, the invariant of `high` is kept while the invariant of `low` is violated once we leave the loop. What we want is the invariant of `low` to be kept and the one of `high` to be violated. How do we achieve that?

The invariant of `high` is never violated because we always make sure that we have *found* a concrete commit that has this `commit_timestamp >= search_timestamp`.

For `low`, we just assign it to `mid + 1` and hope the best. It might have commits that have `commit_timestamp < search_timestamp`, it might also fulfill the criteria for `high`.

What we need to do is to only set `low` if we actually find a `commit_timestamp < search_timestamp`. We can do *that* by inverting `is_latest_commit_timestamp_ge_than` to `is_latest_commit_timestamp_le_than`.

Reply:
(this solution involves setting the `low` only to values where we have run the check for). Obviously, as `mid` rounds down, we cannot do stuff like setting low to `mid` and high to `mid - 1` or such, that might create situations where our output low is higher than the high.

Instead, we just run the check for `mid + 1` instead. Doing that on its own without inverting the check would already ensure that the invariants for `low` don't get invalidated, but we'd still have wasteful WAL to process, so for that we also invert the function's check.